File: web/src/api/insightsClient.test.ts
import { describe, expect, it, vi } from "vitest"

import { InsightsClient } from "@/api/insightsClient"
import type { TokenStore } from "@/api/tokenStore"
import type { InsightsResponse } from "@/api/types"

function memoryStore(token: string | null = "tok"): TokenStore {
  let stored = token
  return {
    get: () => stored,
    set: (value) => {
      stored = value
    },
    clear: () => {
      stored = null
    },
  }
}

const sampleInsights: InsightsResponse = {
  completedSessionCount: 3,
  ready: true,
  ateEnoughYes: 4,
  ateEnoughNo: 2,
  likedLike: 5,
  likedSoSo: 1,
  likedNo: 0,
  likedSkipped: 0,
  topLikedTextures: ["crunchy", "soft"],
  familiarityLikes: 4,
  familiarityFamiliarButNew: 2,
  familiarityTrulyNew: 0,
  snackCount: 1,
  hasParentNotes: false,
  tips: [
    {
      id: "mix_familiarity",
      message: "You've stuck to known foods — when you're ready, try one gentle familiar-but-new.",
    },
    {
      id: "keep_going",
      message: "You're building a tasting rhythm — keep going at a calm pace.",
    },
  ],
}

describe("InsightsClient", () => {
  it("gets insights with bearer auth", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify(sampleInsights), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new InsightsClient(
      "http://localhost:8080",
      fetchFn,
      memoryStore(),
    )
    const insights = await client.get()

    expect(insights.ready).toBe(true)
    expect(insights.tips).toHaveLength(2)
    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/insights",
    )
    const init = fetchFn.mock.calls[0]?.[1] as RequestInit
    expect(new Headers(init.headers).get("Authorization")).toBe("Bearer tok")
  })

  it("dismisses a tip by id", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ status: "ok" }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new InsightsClient(
      "http://localhost:8080",
      fetchFn,
      memoryStore(),
    )
    await client.dismissTip("keep_going")

    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/insights/tips/keep_going/dismiss",
    )
    const init = fetchFn.mock.calls[0]?.[1] as RequestInit
    expect(init.method).toBe("POST")
  })

  it("surfaces dismiss errors for unknown tip ids", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ message: "Unknown tip id: nope" }), {
        status: 400,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new InsightsClient(
      "http://localhost:8080",
      fetchFn,
      memoryStore(),
    )

    await expect(client.dismissTip("nope")).rejects.toThrow(
      "Unknown tip id: nope",
    )
  })
})

File: mobile/sharedLogic/src/commonTest/kotlin/org/example/project/InsightsClientTest.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.serialization.kotlinx.json.json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json

class InsightsClientTest {

    @Test
    fun getSendsBearerAndParsesInsights() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            var sawAuth: String? = null
            val engine =
                MockEngine { request ->
                    assertEquals(HttpMethod.Get, request.method)
                    assertEquals("/api/insights", request.url.encodedPath)
                    sawAuth = request.headers[HttpHeaders.Authorization]
                    respond(
                        content = sampleInsightsJson(),
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = InsightsClient("http://localhost:8080", httpClient(engine), store)
            val insights = client.get()

            assertEquals(3, insights.completedSessionCount)
            assertTrue(insights.ready)
            assertEquals(4, insights.familiaritySafe)
            assertEquals(2, insights.tips.size)
            assertEquals("keep_going", insights.tips[1].id)
            assertEquals("Bearer tok", sawAuth)
        }

    @Test
    fun dismissTipPostsAndParsesOk() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine { request ->
                    assertEquals(HttpMethod.Post, request.method)
                    assertEquals(
                        "/api/insights/tips/keep_going/dismiss",
                        request.url.encodedPath,
                    )
                    respond(
                        content = """{"status":"ok"}""",
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = InsightsClient("http://localhost:8080", httpClient(engine), store)
            val result = client.dismissTip("keep_going")

            assertEquals("ok", result.status)
        }

    @Test
    fun dismissUnknownTipSurfacesMessage() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine {
                    respond(
                        content = """{"message":"Unknown tip id: nope"}""",
                        status = HttpStatusCode.BadRequest,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = InsightsClient("http://localhost:8080", httpClient(engine), store)
            val error =
                assertFailsWith<InsightsException> {
                    client.dismissTip("nope")
                }
            assertEquals("Unknown tip id: nope", error.message)
        }

    private fun httpClient(engine: MockEngine): HttpClient =
        HttpClient(engine) {
            install(ContentNegotiation) {
                json(Json { ignoreUnknownKeys = true })
            }
        }

    private fun sampleInsightsJson(): String =
        """
        {
          "completedSessionCount": 3,
          "ready": true,
          "ateEnoughYes": 4,
          "ateEnoughNo": 2,
          "likedLike": 5,
          "likedSoSo": 1,
          "likedNo": 0,
          "likedSkipped": 0,
          "topLikedTextures": ["crunchy", "soft"],
          "familiaritySafe": 4,
          "familiarityFamiliarButNew": 2,
          "familiarityTrulyNew": 0,
          "snackCount": 1,
          "hasParentNotes": false,
          "tips": [
            {
              "id": "mix_familiarity",
              "message": "You've stuck to known foods — when you're ready, try one gentle familiar-but-new."
            },
            {
              "id": "keep_going",
              "message": "You're building a tasting rhythm — keep going at a calm pace."
            }
          ]
        }
        """.trimIndent()
}
